File: merge_sort/merge_sort.c
// Merge Sort
#include<stdio.h> 

#define SIZE 10

void combine(int *arr,int left, int mid, int right) { 
    int leftP = left;
    int rightP = mid+1;
    int tmpP = left;
    int tmp[SIZE];

    // combining left part and right part in sorted order
    while (leftP <= mid && rightP <= right) {
        if (arr[leftP] <= arr[rightP]) {
            tmp[tmpP++] = arr[leftP++]; // item at leftP goes to tmp arr
        } else {
            tmp[tmpP++] = arr[rightP++];
        }
    }
    // move the remaining
    while (leftP <= mid) {
        tmp[tmpP++] = arr[leftP++];
    }
    while (rightP <= right) {
        tmp[tmpP++] = arr[rightP++];
    }

    // move items at tmp back to arr
    for(int i=left; i<=right; i++){
        arr[i] = tmp[i];
    }
} 

void divide(int *arr, int left, int right) 
{ 
    int mid; 

    if(left < right) { 
        mid = (left+right)/2; 
        divide(arr, left, mid); 
        divide(arr, mid+1, right); 

        combine(arr, left, mid, right); 
    } 
}

int main() { 
    int arr[] = {10,-3,7,15,90,8,0,-5};

    divide(arr,0,SIZE-1); 

    for(int i = 0; i<SIZE; i++) {
        printf("%d ",arr[i]); 
    }
    return 0;
} 

File: merge_sort/mergesort.py
# the step of combining sorted subarrays tgt
def merge(left_sorted_array, right_sorted_array):
    p1, p2 = 0, 0
    merged_array = []

    # combining 2 sorted arrays into 1
    while (p1 < len(left_sorted_array) and p2 < len(right_sorted_array)):
        if left_sorted_array[p1] < right_sorted_array[p2]:
            merged_array.append(left_sorted_array[p1])
            p1 += 1
        else:
            merged_array.append(right_sorted_array[p2])
            p2 += 1
    
    # if there's anything left in either array
    merged_array.extend(left_sorted_array[p1:])
    merged_array.extend(right_sorted_array[p2:])
    return merged_array

def mergesort(array):
    if len(array) <= 1:
        return array

    # sort subarrays
    mid = len(array) // 2
    left_sorted_array = mergesort(array[:mid])
    right_sorted_array = mergesort(array[mid:])

    # merge left & right sorted arrays
    return merge(left_sorted_array, right_sorted_array)


if __name__ == "__main__":
    array = [12, 11, 13, 5, 6, 7]
    print("unsorted array:")
    print(array)

    sorted_array = mergesort(array)
    print("sorted array:")
    print(sorted_array)